### PR TITLE
[ORCA-163] Implement `launch_workflow()` and `LaunchInfo` dataclass

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
I was initially going to port the `launch_workflow()` implementation from sagetasks ([source](https://github.com/Sage-Bionetworks-Workflows/sagetasks/blob/524b5112a2d7e6b083304e0528fa8d7876f8601a/src/sagetasks/nextflowtower/utils.py#L189-L272)), but I decided to refactor it. Its reliance on the `init_launch_workflow_data()` method is clunky. I also didn't like the large number of function parameters. 

To address this, I implemented the `LaunchInfo` dataclass. Its goal is to encapsulate the information related to launching a workflow. It essentially operates like an OpenAPI model. This results in much cleaner client code. This class can eventually be reused for `get_workflow()`. 

I'm marking this PR as a draft because I would like to tackle the following subtasks:

- [x] Implement `get_compute_env()` in the client class
- [x] Implement `get_latest_compute_env()` in the ops class
- [x] Implement higher-level `launch_workflow()` in the ops class
  - This version would use the default values from the compute environment (_e.g._ work directory) when launching a workflow ([example](https://github.com/Sage-Bionetworks-Workflows/sagetasks/blob/524b5112a2d7e6b083304e0528fa8d7876f8601a/src/sagetasks/nextflowtower/utils.py#L156-L159))
  - This version would ideally be idempotent, if possible

### CODE REVIEW

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py:129]: Is this the response people get when they call: `dataclasses.asdict(obj)`?

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py:1]: Thanks for doing this - it makes the code a lot cleaner!  Not only can we pass this model around and but it also adds a layer of documentation for the request object without going to the API.

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py]: Nit: https://docs.pydantic.dev/usage/models/#helper-functions.  Would this be helpful? (in particular the `parse_obj` function included in `BaseModels`)

That said, that is technically beyond the scope of this PR.

Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py]: That doesn't work for most models because the API responses use camel case and the data classes use snake case. We could make this work if we used camel case for our data classes. Not sure if we want to break convention though.

File: src/orca/services/nextflowtower/models.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py:129]: See my comment [here](https://github.com/Sage-Bionetworks-Workflows/py-orca/pull/15#discussion_r1180847295).

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [reply src/orca/services/nextflowtower/models.py]: Ya... let's not break the convention. Was just curious if we could reuse something.

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py:115]: What is your rationale of returning a list rather than a generator?

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py]: I appreciate your intention for saving one rest call, but what are your thoughts on the added complexity of the code over the one rest call that you save?

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py:145]: Is there a need to add support for pagination here?

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py:289]: Just annotating - this is ORCA-205. Please self assign and pull into sprint.

File: src/orca/services/nextflowtower/ops.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/ops.py:51]: Just annotating - this is ORCA-166. Please self assign and pull into sprint.

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py:91]: Just a note, we may want to take a look at the [request body](https://tower.nf/openapi/index.html#post-/workflow/launch) and decide which ones we want to add in the future (you have the request body below as well). For example, "resume" might be a good one, but out of scope for this ticket.

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py]: This is interesting, I didn't realize `dateCreated` was not a server-filled attribute.  Just curious - what are your thoughts on this sort of architecture - leaving it to the clients to fill out this information.

File: src/orca/services/nextflowtower/models.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/models.py]: Should there be a `from_response` dict here?  That said, the response object of a workflow launch is vastly different, so it might not make any sense.

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py]: Just a note: I have to keep reminding myself that `response` is not actually a `requests.Response` object, but it's the json.

File: tests/services/nextflowtower/responses.py
Comment (thomasyu888) [inline tests/services/nextflowtower/responses.py:157]: This isn't isoformat which is causing a test to fail:
```
datetime.fromisoformat("2023-04-26T00:49:49Z")
# ValueError: Invalid isoformat string: '2023-04-26T00:49:49Z'
datetime.fromisoformat("2023-04-26T00:49:49")
# datetime.datetime(2023, 4, 26, 0, 49, 49)
```

File: tests/services/nextflowtower/test_ops.py
Comment (thomasyu888) [inline tests/services/nextflowtower/test_ops.py:68]: Same thing here with the isoformat strings
```suggestion
        generate_compute_env("5ykJF", "2023-04-26T00:49:49"),
        generate_compute_env("9W2l7", "2023-01-11T13:32:54"),
```

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py:269]: I could be wrong about this, but I'm not sure everyone actually has the ability to create labels.  It may be problematic if it is an entirely new workspace and someone tries to create the `launched-by-orca` label when they leverage this package to launch workflows.

File: src/orca/services/nextflowtower/ops.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/ops.py:96]: I'm not fully understanding here, why is this a `continue`?  Even if we pass in a label that is a `resource`, don't we want to return the `label.id`?

This test: `test_that_create_label_ignores_resource_labels` is what spurs the discussion.  Should `create_label` be called if the resource label exists?

**Review by thomasyu888 — CHANGES_REQUESTED** [review]
Awesome!  Thanks for putting this together.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py:289]: Done!

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py:269]: I've been able to create labels with my personal user, which has the "launcher" role. This would be the minimum level of access needed by Orca if you want to launch workflows. If you try to launch a workflow with a read-only token, you should run into an unauthorized exception.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py]: I agree that it's a little confusing. Using `json` as the variable name should be clearer. I'll update.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py:145]: Not yet. I don't know if there are any paginated POST endpoints. As far as I know, pagination is only used for GET endpoints that return long lists of resources, whereas POST requests usually return info about the resource you just created.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py]: Fair enough. I'll simplify the code at the cost of introducing an extra HTTP request.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py:115]: It's a more consistent interface between paginated and non-paginated requests. If memory usage becomes an issue, we can revisit this.

File: src/orca/services/nextflowtower/models.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py]: YAGNI. I try to avoid implementing methods until I need them. We might be able to use this model once we merge this branch with Brad's `get_workflow()` PR.

File: src/orca/services/nextflowtower/models.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py]: I'm emulating the behavior that I'm seeing with the Tower web client. I'll try to launch a workflow without this field. If I can, then I'll drop this because it's simpler. IMHO, this feels like something that the backend should handle.

Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py]: I was able to run the `test_that_a_workflow_can_be_launched` integration test while setting this field to None, so I'm going to remove the timestamp code.

File: src/orca/services/nextflowtower/client.py
Comment (BWMac) [inline src/orca/services/nextflowtower/client.py]: Might it make sense to have functions like this one in a cross-service `utils` file? I can imagine this being useful for any service interacting with an API.

File: src/orca/services/nextflowtower/models.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/models.py:91]: You captured my intent with these models. Some of the models in the OpenAPI spec are very big. Replicating them here feels unnecessary, especially if we only use a small subset of the fields. As we need to add new features, we can expand what these models can store. 

I attempted a compromise with the `ComputeEnv` model by having a `raw` attribute that stores the full JSON response, so we'll see if that approach works well.

File: src/orca/services/nextflowtower/ops.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/ops.py:96]: I'm trying to differentiate between workflow labels and resource tags, which Tower lumps together as labels. This function is intended to be about workflow labels. I updated the docstring to clarify. 

Note that a (non-resource) workflow label can coexist with a resource tag with the same name. Here, I was able to create a `CostCenter` label despite a `CostCenter` resource tag already existing:

```json
{
  "id": 89366116242820,
  "name": "CostCenter",
  "value": "000000",
  "resource": true
}

{
  "id": 266513533567473,
  "name": "CostCenter",
  "value": null,
  "resource": false
}
```

File: src/orca/services/nextflowtower/ops.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/ops.py:51]: Done!

File: tests/services/nextflowtower/responses.py
Comment (BrunoGrandePhD) [reply tests/services/nextflowtower/responses.py:157]: Interesting. This works fine on my machine:

```python
>>> from datetime import datetime
>>> datetime.fromisoformat("2023-04-26T00:49:49Z")
datetime.datetime(2023, 4, 26, 0, 49, 49, tzinfo=datetime.timezone.utc)
>>> datetime.fromisoformat("2023-04-26T00:49:49")
datetime.datetime(2023, 4, 26, 0, 49, 49)
```

I guess the issue is that this is technically RFC 3339 rather than ISO 8601. I implemented a simple parser utility function to handle these RFC 3339 timestamps from Tower. If this proves to be tricky going forward, we can look into using `pendulum` instead, which is already installed as a dependency of Airflow. I'm holding off for now as the current approach is more strict (I want to know if the format is consistent), and the `pendulum.parse()` function doesn't return a consistent type, so the code will be slightly more cluttered to avoid typing errors.

File: tests/services/nextflowtower/test_ops.py
Comment (BrunoGrandePhD) [reply tests/services/nextflowtower/test_ops.py:68]: I updated the datetime parsing so that I can consistently use the same format adopted by Tower, which follows RFC 3339 (not ISO 8601).

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py]: I agree. Once we add a second service where we need to implement a client (like Tower), we can abstract the common functionality to a base client class. Fortunately, most services that we do plan on using have clients (_e.g._ Synapse, SevenBridges). We might need to reuse this code once we implement support for GA4GH WES, but I don't think that's prioritized for the time being.

**Comment by BrunoGrandePhD** [comment]
Thank goodness for the test suite! I was able to resolve multiple merge conflicts with confidence.

File: src/orca/services/nextflowtower/client.py
Comment (thomasyu888) [inline src/orca/services/nextflowtower/client.py:111]: Nit: `json` as a variable conflicts with the base `json` package. It might be fine since we aren't importing it, but if we did decide to, this variable name might cause problems. Maybe `response_json` (I know it's long...)

File: src/orca/services/nextflowtower/ops.py
Comment (thomasyu888) [reply src/orca/services/nextflowtower/ops.py:96]: Thanks for the clarification - I did not realize you could have resource and non-resource labels that are named the same.

**Review by thomasyu888 — APPROVED** [review]
🛸 Excellent!  Just a minor comment about the `json` variable, but going to pre-approve.

File: src/orca/services/nextflowtower/client.py
Comment (BrunoGrandePhD) [reply src/orca/services/nextflowtower/client.py:111]: Ah, that's probably why I didn't opt for `json` or `dict` at first. Let's stick with this for now. It's possible that we can avoid importing the `json` module where the `json` variable exists.

**Pull request merged**
